src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
+      let path = path ~root ~chunk_idx in
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      (* Passing [end_poff] is required by Ao for its consistency check (and
+         only applies to the appendable chunk).
+         TODO: We could change its API in the future so we can not do
+         [Io.size_of_path], which is admittedly very silly.*)
+      let* end_poff =
+        if is_appendable then Ok end_poff else Io.size_of_path path
+      in
+      Ao.open_ro ~path ~end_poff ~dead_header_size
+  end
+
+  let create_rw ~root ~start_idx ~overwrite ~auto_flush_threshold
+      ~auto_flush_procedure =
+    let open Result_syntax in
+    let chunk_idx = start_idx in
+    let+ ao =
+      Ao_chunk.create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure
+    in
+    let chunk = { idx = chunk_idx; start_off = Int63.zero; ao } in
+    let inventory = Inventory.v 1 (Fun.const chunk) in
+    { inventory }
+
+  let open_rw ~root ~end_poff ~start_idx ~chunk_num ~dead_header_size
+      ~auto_flush_threshold ~auto_flush_procedure =
+    let open Result_syntax in
+    let open_chunk ~chunk_idx ~is_appendable =
+      match is_appendable with
+      | true ->
+          Ao_chunk.open_rw ~root ~chunk_idx ~end_poff ~auto_flush_threshold
+            ~auto_flush_procedure ~dead_header_size
+      | false ->
+          Ao_chunk.open_ro ~root ~chunk_idx ~end_poff ~dead_header_size
+            ~is_appendable
+    in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let open_ro ~root ~end_poff ~dead_header_size ~start_idx ~chunk_num =
+    let open Result_syntax in
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
+
+  let read_exn t ~off ~len buf =
+    let chunk = Inventory.find ~off t.inventory in
+    let poff = chunk_off_to_poff chunk off in
+    Ao.read_exn chunk.ao ~off:poff ~len buf
+
+  let append_exn t s = Ao.append_exn (appendable_ao t) s
+
+  let close t =
+    (* Close immutable chunks, ignoring errors. *)
+    let _ =
+      t.inventory
+      |> Inventory.iter_skip_appendable @@ fun chunk ->
+         let _ = Ao.close chunk.ao in
+         ()
+    in
+    (* Close appendable chunk and keep error since this
+       is the one that can have a pending flush. *)
+    appendable_ao t |> Ao.close
```

Comment:
Nitpick: Perhaps it would be easier to implement the `close` in the `Inventory` module rather than expose an `iter_skip_appendable`?

Reply:
I iterated a bit on the boundary between `Inventory` and the encompassing module. The current distinction that I settled upon is that `Inventory` is for storing/looking up chunks (with light concepts like appendable for easier access) but doesn't call functions on `Ao` (with the exception of `end_poff` since that is a property used for finding chunks). I wanted any interactions with the `Ao.t` to be in the domain of the encompassing module.

I am definitely open to exploring this boundary though! Maybe see how subsequent changes go and revise later?

Reply:
@Ngoguey42 okay, that discrepancy convinces me to move it. I cannot unsee that now.

Reply:
[Made the change](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825)!